chainerx_cc/chainerx/cuda/cuda_device/indexing.cu

```diff
@@ -185,58 +170,57 @@ void TakeImpl(Device& device, const Array& a, const Array& indices, int8_t axis,
-        Axes a_perm = MakeRollingPermutation(axis, axis + 1, a.ndim());
-        a_iarray.Permute(a_perm);
-        Shape a_shape = internal::TransposeShape(a.shape(), a_perm);
-        Indexer<> a_indexer{a_shape};
-
-        IndexableArray<T> out_iarray{out};
-        Axes out_perm = MakeRollingPermutation(axis, axis + indices.ndim(), out.ndim());
-        out_iarray.Permute(out_perm);
-        Shape out_shape = internal::TransposeShape(out.shape(), out_perm);
-        Indexer<> out_indexer{out_shape};
-
-        IndexableArray<const TIndex> indices_iarray{indices};
-        Indexer<> indices_indexer{indices.shape()};
-
-        // size of (Ni..., Nj...) part
-        TIndex common_total_size = gsl::narrow<TIndex>(a_indexer.total_size() / a_shape[0]);
-
-        TIndex axis_dim = gsl::narrow<TIndex>(a_shape[0]);
-
-        // TODO(niboshi): Calculate kMaxBlockSize per device
-        std::lock_guard<std::mutex> lock{*cuda_internal::g_mutex};
-        int64_t total_size = out_indexer.total_size();
-        static const int kWrapMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeWrapCudaKernel<T, TIndex>).block_size;
-        static const int kClipMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeClipCudaKernel<T, TIndex>).block_size;
-        int64_t grid_size;
-        int64_t block_size;
-        switch (mode) {
-            case IndexBoundsMode::kRaise:
-                throw BackendError{"Take with mode='raise' is not supported with CUDA backend"};
-            case IndexBoundsMode::kDefault:
-            case IndexBoundsMode::kWrap:
-                grid_size = (total_size + kWrapMaxBlockSize - 1) / kWrapMaxBlockSize;
-                block_size = std::min<TIndex>(total_size, kWrapMaxBlockSize);
-                TakeWrapCudaKernel<<<grid_size, block_size>>>(
-                        a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
-                break;
-            case IndexBoundsMode::kClip:
-                grid_size = (total_size + kClipMaxBlockSize - 1) / kClipMaxBlockSize;
-                block_size = std::min<TIndex>(total_size, kClipMaxBlockSize);
-                TakeClipCudaKernel<<<grid_size, block_size>>>(
-                        a_iarray, out_iarray, indices_iarray, a_indexer, out_indexer, indices_indexer, common_total_size, axis_dim);
-                break;
+        int64_t num_indices = indices.GetTotalSize();
+        int64_t target_dim = a.shape()[axis];
+        int64_t num_iters = out.GetTotalSize();
+        int64_t right_dim = std::accumulate(a.shape().begin() + axis + 1, a.shape().end(), int64_t{1}, std::multiplies<>());
+        static const int k1DMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeCudaKernel<T, TIndex, 1>).block_size;
+        static const int kNDMaxBlockSize = CudaOccupancyMaxPotentialBlockSize(&TakeCudaKernel<T, TIndex, kDynamicNdim>).block_size;
+        if (a.IsContiguous() && indices.IsContiguous()) {
```

Comment:
Why did you require `indices` to be contiguous? I have not compared the performance yet, but it looks not performance-critical.

Reply:
Even if the input is not contiguous, the operation sometimes can get speed up.

1. Sometimes the continuous two axes can be fused and reduced the ndim.
1. `IndexableArray<const T, 1>` is available when `a` is non-contiguous `1`-dim array.

How about implementing something like `internal::ReduceContiguousDims(Array)`?

Reply:
`indices` is required to be contiguous so we can use the IndexableArray<TIndex, 1> instead of the dynamic one.
In my initial experiments if `indices` was not a 1D array,performance was 3 times worse.
Maybe we can apply the second suggestion?

Reply:
So we allow to use non-contiguous arrays (with strides) as 1D arrays.
And try to reduce dimensionality of ND arrays when creating the Indexable arrays to broad the scope of cases that can be accelerated.
This seems promising!